### ISSUE TITLE
Native elisp load failed on Apple M1: mach-o file, but is an incompatible architecture (have (arm64), need (x86_64))

### ISSUE THREAD

**Issue by goofansu** [doomemacs/doomemacs#7228]
### I confirm that...

- [X] I have searched the [issue tracker](https://github.com/doomemacs/doomemacs/issues), [documentation](https://docs.doomemacs.org), [FAQ](https://docs.doomemacs.org/-/faq), [Discourse](https://discourse.doomemacs.org), and [Google](https://google.com), in case this issue has already been reported/resolved.
- [X] I have read "[How to Debug Issues](https://doomemacs.org/d/how2debug)", and will use it to provide as much information about this issue as possible.
- [X] The issue can be reproduced on the **latest** available commit of Doom.
- [X] The issue can be reproduced on a stable release of Emacs, such as 27 or 28. *(Doom does not support development builds like 29+ or any version ending in .50 or .9x)*

### Expected behavior

I'm using Apple M1 (macOS 12.6) with a x86 Emacs.app (GNU Emacs 28.2 (build 1, x86_64-apple-darwin20.6.0, NS appkit-2202.70 Version 11.6.8 (Build 20G730)) of 2022-09-13) from https://github.com/jimeh/emacs-builds.

When visiting a directory with `direnv` enabled, if packages are not native-compiled before, they should be compiled in the x86 architecture to be loaded correctly.

### Current behavior

Some packages are compiled in the arm64 architecture.

The message buffer:

```
Loading /Users/james/.emacs.d/.custom.el (source)...done
Doom loaded 265 packages across 55 modules in 3.038s
Running direnv in /Users/james/nobackup/emacs-nativecomp-error-in-direnv/ ... (C-g to abort)
Direnv succeeded in /Users/james/nobackup/emacs-nativecomp-error-in-direnv/
Loading /Users/james/.emacs.d/.local/cache/recentf...done
error in process sentinel: Native elisp load failed: "/Users/james/.emacs.d/.local/cache/eln/28_2-7e51a3b1/dtrt-indent-781f5f33-fc9709f7.eln", "dlopen(/Users/james/.emacs.d/.local/cache/eln/28_2-7e51a3b1/dtrt-indent-781f5f33-fc9709f7.eln, 0x0001): tried: '/Users/james/.emacs.d/.local/cache/eln/28_2-7e51a3b1/dtrt-indent-781f5f33-fc9709f7.eln' (mach-o file, but is an incompatible architecture (have (arm64), need (x86_64)))" [2 times]
error in process sentinel: Native elisp load failed: "/Users/james/.emacs.d/.local/cache/eln/28_2-7e51a3b1/use-package-core-38e7bdb9-a8e5f021.eln", "dlopen(/Users/james/.emacs.d/.local/cache/eln/28_2-7e51a3b1/use-package-core-38e7bdb9-a8e5f021.eln, 0x0001): tried: '/Users/james/.emacs.d/.local/cache/eln/28_2-7e51a3b1/use-package-core-38e7bdb9-a8e5f021.eln' (mach-o file, but is an incompatible architecture (have (arm64), need (x86_64)))" [2 times]
error in process sentinel: Native elisp load failed: "/Users/james/.emacs.d/.local/cache/eln/28_2-7e51a3b1/bind-key-432e7b9b-dd3bfcce.eln", "dlopen(/Users/james/.emacs.d/.local/cache/eln/28_2-7e51a3b1/bind-key-432e7b9b-dd3bfcce.eln, 0x0001): tried: '/Users/james/.emacs.d/.local/cache/eln/28_2-7e51a3b1/bind-key-432e7b9b-dd3bfcce.eln' (mach-o file, but is an incompatible architecture (have (arm64), need (x86_64)))" [2 times]
error in process sentinel: Native elisp load failed: "/Users/james/.emacs.d/.local/cache/eln/28_2-7e51a3b1/evil-vars-2cd7d1a0-d9108f00.eln", "dlopen(/Users/james/.emacs.d/.local/cache/eln/28_2-7e51a3b1/evil-vars-2cd7d1a0-d9108f00.eln, 0x0001): tried: '/Users/james/.emacs.d/.local/cache/eln/28_2-7e51a3b1/evil-vars-2cd7d1a0-d9108f00.eln' (mach-o file, but is an incompatible architecture (have (arm64), need (x86_64)))" [2 times]
Mark set [3 times]
Blocking call to accept-process-output with quit inhibited!!
```

All eln list:
```
> ls | xargs -I {} file {}
all-the-icons-0a2d9f0b-e618b6a0.eln: Mach-O 64-bit dynamically linked shared library x86_64
annalist-952cc743-bf925ab7.eln: Mach-O 64-bit dynamically linked shared library x86_64
async-34298af2-69cfd031.eln: Mach-O 64-bit dynamically linked shared library x86_64
better-jumper-6af02a75-2ab834b2.eln: Mach-O 64-bit dynamically linked shared library x86_64
bind-key-432e7b9b-dd3bfcce.eln: Mach-O 64-bit dynamically linked shared library arm64
company-2905c617-96d34bde.eln: Mach-O 64-bit dynamically linked shared library x86_64
company-box-88dc31c4-59fdef2a.eln: Mach-O 64-bit dynamically linked shared library x86_64
company-box-doc-aa96eeb4-854b1d24.eln: Mach-O 64-bit dynamically linked shared library x86_64
company-box-icons-bbd7844f-da3d0887.eln: Mach-O 64-bit dynamically linked shared library x86_64
company-capf-fdf8d9ac-97f64634.eln: Mach-O 64-bit dynamically linked shared library x86_64
compat-29-16d2264e-68403704.eln: Mach-O 64-bit dynamically linked shared library x86_64
consult-9b2e4cd1-29487f2c.eln: Mach-O 64-bit dynamically linked shared library x86_64
consult-flycheck-a4d15586-82862628.eln: Mach-O 64-bit dynamically linked shared library x86_64
consult-vertico-f9ab7a13-0b9a289e.eln: Mach-O 64-bit dynamically linked shared library x86_64
dash-b27d142c-1c8b1bbd.eln: Mach-O 64-bit dynamically linked shared library x86_64
diredfl-a6f05875-88c3ef1e.eln: Mach-O 64-bit dynamically linked shared library x86_64
doom-modeline-afc2615c-0a15a74d.eln: Mach-O 64-bit dynamically linked shared library x86_64
doom-modeline-core-4d2a5d12-dcf952a0.eln: Mach-O 64-bit dynamically linked shared library x86_64
doom-modeline-env-b68d0d9e-40c64712.eln: Mach-O 64-bit dynamically linked shared library x86_64
doom-modeline-segments-c3876916-3d307aa2.eln: Mach-O 64-bit dynamically linked shared library x86_64
doom-snippets-c40c3c5f-b7a3035a.eln: Mach-O 64-bit dynamically linked shared library x86_64
doom-snippets-lib-fa5c05b3-4473d159.eln: Mach-O 64-bit dynamically linked shared library x86_64
doom-themes-d73167e4-7d56fb77.eln: Mach-O 64-bit dynamically linked shared library x86_64
doom-themes-ext-org-e6d3664b-5822900a.eln: Mach-O 64-bit dynamically linked shared library x86_64
dtrt-indent-781f5f33-fc9709f7.eln: Mach-O 64-bit dynamically linked shared library arm64
emacsql-b23e90a6-a4241958.eln: Mach-O 64-bit dynamically linked shared library x86_64
emacsql-compiler-0ed49d41-3de5951f.eln: Mach-O 64-bit dynamically linked shared library x86_64
emacsql-sqlite-53ef6591-913e2b28.eln: Mach-O 64-bit dynamically linked shared library x86_64
emacsql-sqlite-builtin-7e2fcd87-cee7dbe7.eln: Mach-O 64-bit dynamically linked shared library x86_64
emacsql-sqlite-common-6c13f7fb-c48e95e5.eln: Mach-O 64-bit dynamically linked shared library x86_64
emojify-8795b470-8c4fd78f.eln: Mach-O 64-bit dynamically linked shared library x86_64
envrc-ae29e37b-070a21db.eln: Mach-O 64-bit dynamically linked shared library x86_64
evil-mc-common-7ec46333-dbc80254.eln: Mach-O 64-bit dynamically linked shared library x86_64
evil-mc-cursor-state-3c237490-c82f2877.eln: Mach-O 64-bit dynamically linked shared library x86_64
evil-mc-region-3a7b62be-2bcf4614.eln: Mach-O 64-bit dynamically linked shared library x86_64
evil-vars-2cd7d1a0-d9108f00.eln: Mach-O 64-bit dynamically linked shared library arm64
f-ce28dbe1-24f45ff3.eln: Mach-O 64-bit dynamically linked shared library x86_64
flycheck-f09a1cdc-810ebd28.eln: Mach-O 64-bit dynamically linked shared library x86_64
flycheck-package-fd85f94d-1c268aa0.eln: Mach-O 64-bit dynamically linked shared library x86_64
flycheck-popup-tip-b2343559-451bf160.eln: Mach-O 64-bit dynamically linked shared library x86_64
frame-local-ef4a464f-63e12a98.eln: Mach-O 64-bit dynamically linked shared library x86_64
gcmh-b95c539e-70d70e65.eln: Mach-O 64-bit dynamically linked shared library x86_64
general-f792a3ea-2cc9dbec.eln: Mach-O 64-bit dynamically linked shared library x86_64
git-commit-a472cede-d24f7915.eln: Mach-O 64-bit dynamically linked shared library x86_64
hide-mode-line-f55f7629-cc4a797c.eln: Mach-O 64-bit dynamically linked shared library x86_64
hl-todo-58f4ab8e-e4ced0fb.eln: Mach-O 64-bit dynamically linked shared library x86_64
ht-837969cc-8ba98b75.eln: Mach-O 64-bit dynamically linked shared library x86_64
ibuffer-vc-b0d7f840-2e2cfede.eln: Mach-O 64-bit dynamically linked shared library x86_64
inheritenv-49bb96ba-262ab9fb.eln: Mach-O 64-bit dynamically linked shared library x86_64
magit-apply-6e1f406c-5a131e5c.eln: Mach-O 64-bit dynamically linked shared library x86_64
magit-autorevert-c1e98083-e0c92f8c.eln: Mach-O 64-bit dynamically linked shared library x86_64
magit-base-a58efc05-90841f01.eln: Mach-O 64-bit dynamically linked shared library x86_64
magit-bisect-95626ac1-2a354cbb.eln: Mach-O 64-bit dynamically linked shared library x86_64
magit-blame-facff7da-606bb481.eln: Mach-O 64-bit dynamically linked shared library x86_64
magit-bookmark-a96bb687-b2c9e2a9.eln: Mach-O 64-bit dynamically linked shared library x86_64
magit-branch-1c4667dd-598fe16b.eln: Mach-O 64-bit dynamically linked shared library x86_64
magit-clone-215a52e9-71ff2732.eln: Mach-O 64-bit dynamically linked shared library x86_64
magit-commit-ade51b95-fd608f97.eln: Mach-O 64-bit dynamically linked shared library x86_64
magit-diff-895cf8f0-6e776189.eln: Mach-O 64-bit dynamically linked shared library x86_64
magit-ee1afbaa-d7adfeec.eln: Mach-O 64-bit dynamically linked shared library x86_64
magit-fetch-56fcb970-40bac797.eln: Mach-O 64-bit dynamically linked shared library x86_64
magit-files-18267077-1c55a78f.eln: Mach-O 64-bit dynamically linked shared library x86_64
magit-git-cbb67f39-03e896bb.eln: Mach-O 64-bit dynamically linked shared library x86_64
magit-log-cf3e40cc-1c738c9b.eln: Mach-O 64-bit dynamically linked shared library x86_64
magit-margin-96bb3d3e-4b1a9cf2.eln: Mach-O 64-bit dynamically linked shared library x86_64
magit-merge-c41d2c23-5dff1c6e.eln: Mach-O 64-bit dynamically linked shared library x86_64
magit-mode-9a659f53-4e7ff2b7.eln: Mach-O 64-bit dynamically linked shared library x86_64
magit-notes-0d09f74b-834acaf8.eln: Mach-O 64-bit dynamically linked shared library x86_64
magit-obsolete-41881bb9-2b0ae72a.eln: Mach-O 64-bit dynamically linked shared library x86_64
magit-process-a25799b9-958a1d13.eln: Mach-O 64-bit dynamically linked shared library x86_64
magit-pull-67c66bf6-3bf3af7f.eln: Mach-O 64-bit dynamically linked shared library x86_64
magit-push-035031fd-1462a17e.eln: Mach-O 64-bit dynamically linked shared library x86_64
magit-reflog-a13adf96-69fc9edb.eln: Mach-O 64-bit dynamically linked shared library x86_64
magit-refs-3073fb74-60a16485.eln: Mach-O 64-bit dynamically linked shared library x86_64
magit-remote-e766e02e-63db2e52.eln: Mach-O 64-bit dynamically linked shared library x86_64
magit-repos-549d7d68-36de3adc.eln: Mach-O 64-bit dynamically linked shared library x86_64
magit-reset-ab47fcb9-709d5f56.eln: Mach-O 64-bit dynamically linked shared library x86_64
magit-section-ecd6b435-8cdf5ed8.eln: Mach-O 64-bit dynamically linked shared library x86_64
magit-sequence-65f12f89-136bd8ad.eln: Mach-O 64-bit dynamically linked shared library x86_64
magit-stash-69ac1883-227dbe3c.eln: Mach-O 64-bit dynamically linked shared library x86_64
magit-status-ef21867d-49d61767.eln: Mach-O 64-bit dynamically linked shared library x86_64
magit-status-ef21867d-49d61767r5fdh1.eln.tmp: empty
magit-submodule-9d66e8ec-515da988.eln: Mach-O 64-bit dynamically linked shared library x86_64
magit-tag-cf3c8bb4-d7afb432.eln: Mach-O 64-bit dynamically linked shared library x86_64
magit-todos-5fac8d35-7ded5fbc.eln: Mach-O 64-bit dynamically linked shared library x86_64
magit-transient-d0c7d5b7-f0633e58.eln: Mach-O 64-bit dynamically linked shared library x86_64
magit-wip-5f2ae370-31bba412.eln: Mach-O 64-bit dynamically linked shared library x86_64
magit-worktree-a1770c34-d20ec892.eln: Mach-O 64-bit dynamically linked shared library x86_64
marginalia-638301b9-00136c0e.eln: Mach-O 64-bit dynamically linked shared library x86_64
ns-auto-titlebar-71b765cc-afaefc1e.eln: Mach-O 64-bit dynamically linked shared library x86_64
ob-comint-738b3c93-08bfe38c.eln: Mach-O 64-bit dynamically linked shared library x86_64
ob-core-663ef2eb-b501abc4.eln: Mach-O 64-bit dynamically linked shared library x86_64
ob-emacs-lisp-8f17239c-1d800b2a.eln: Mach-O 64-bit dynamically linked shared library x86_64
ob-eval-aae63a72-f2f08e88.eln: Mach-O 64-bit dynamically linked shared library x86_64
ob-exp-5b14b801-07d0706f.eln: Mach-O 64-bit dynamically linked shared library x86_64
ob-lob-7c77e7bb-f2b24928.eln: Mach-O 64-bit dynamically linked shared library x86_64
ob-ref-3addabcc-6b8a9ccc.eln: Mach-O 64-bit dynamically linked shared library x86_64
ob-table-f982f02e-a87cc52a.eln: Mach-O 64-bit dynamically linked shared library x86_64
ob-tangle-a68d908a-444eaa19.eln: Mach-O 64-bit dynamically linked shared library x86_64
oc-a0393fec-24f9c028.eln: Mach-O 64-bit dynamically linked shared library x86_64
ol-5d425762-a0d77132.eln: Mach-O 64-bit dynamically linked shared library x86_64
orderless-1ce485f3-ef71cfa6.eln: Mach-O 64-bit dynamically linked shared library x86_64
org-35b93c7d-77ba6e83.eln: Mach-O 64-bit dynamically linked shared library x86_64
org-agenda-968cfa6e-d1f2eee2.eln: Mach-O 64-bit dynamically linked shared library x86_64
org-attach-a0869377-6c6fa892.eln: Mach-O 64-bit dynamically linked shared library x86_64
org-capture-6cfa9962-459e250b.eln: Mach-O 64-bit dynamically linked shared library x86_64
org-compat-d32a4aaa-fbeaf99d.eln: Mach-O 64-bit dynamically linked shared library x86_64
org-cycle-900ebd49-076fcc33.eln: Mach-O 64-bit dynamically linked shared library x86_64
org-element-9a0cb642-0fd70c3f.eln: Mach-O 64-bit dynamically linked shared library x86_64
org-entities-e9229475-15e67312.eln: Mach-O 64-bit dynamically linked shared library x86_64
org-faces-f9ee800b-41ba3366.eln: Mach-O 64-bit dynamically linked shared library x86_64
org-fold-1f2bbb32-429663f0.eln: Mach-O 64-bit dynamically linked shared library x86_64
org-fold-core-1cc39ca5-ee070228.eln: Mach-O 64-bit dynamically linked shared library x86_64
org-footnote-101cc21d-9c6cc18e.eln: Mach-O 64-bit dynamically linked shared library x86_64
org-id-24057503-01eea74f.eln: Mach-O 64-bit dynamically linked shared library x86_64
org-keys-fd497d29-57065d07.eln: Mach-O 64-bit dynamically linked shared library x86_64
org-list-e7904209-5c281cc2.eln: Mach-O 64-bit dynamically linked shared library x86_64
org-macro-0d687d46-a8c1d25a.eln: Mach-O 64-bit dynamically linked shared library x86_64
org-macs-2963cad9-fd9a5918.eln: Mach-O 64-bit dynamically linked shared library x86_64
org-pcomplete-cd1b0499-422d58af.eln: Mach-O 64-bit dynamically linked shared library x86_64
org-persist-623df054-471a52b7.eln: Mach-O 64-bit dynamically linked shared library x86_64
org-refile-55187b21-f6969aca.eln: Mach-O 64-bit dynamically linked shared library x86_64
org-roam-capture-5bffad22-9886c557.eln: Mach-O 64-bit dynamically linked shared library x86_64
org-roam-cfad0ebc-9621a6c9.eln: Mach-O 64-bit dynamically linked shared library x86_64
org-roam-compat-3ee94bdb-aa6ebb31.eln: Mach-O 64-bit dynamically linked shared library x86_64
org-roam-dailies-7d8c718a-1f837264.eln: Mach-O 64-bit dynamically linked shared library x86_64
org-roam-db-4248498a-f7350f11.eln: Mach-O 64-bit dynamically linked shared library x86_64
org-roam-id-c8907bf7-05f26bd7.eln: Mach-O 64-bit dynamically linked shared library x86_64
org-roam-log-d66f95cb-40352623.eln: Mach-O 64-bit dynamically linked shared library x86_64
org-roam-migrate-b944c063-0404f457.eln: Mach-O 64-bit dynamically linked shared library x86_64
org-roam-mode-f62cd8ed-a44fa28c.eln: Mach-O 64-bit dynamically linked shared library x86_64
org-roam-node-10706de3-caa67140.eln: Mach-O 64-bit dynamically linked shared library x86_64
org-roam-utils-c1d6a992-d2ccb307.eln: Mach-O 64-bit dynamically linked shared library x86_64
org-src-8ba3107e-b894e377.eln: Mach-O 64-bit dynamically linked shared library x86_64
org-table-6a578e66-01d4e2a0.eln: Mach-O 64-bit dynamically linked shared library x86_64
org-version-e2d5ff00-220dcec3.eln: Mach-O 64-bit dynamically linked shared library x86_64
org-yt-c947f2d1-73b9a7ba.eln: Mach-O 64-bit dynamically linked shared library x86_64
package-lint-5716469a-c7f10cdf.eln: Mach-O 64-bit dynamically linked shared library x86_64
pcre2el-a6ffbbe2-0ad91567.eln: Mach-O 64-bit dynamically linked shared library x86_64
persp-mode-0fb5c396-7fea9877.eln: Mach-O 64-bit dynamically linked shared library x86_64
popup-d2f1ff21-4dc33e1b.eln: Mach-O 64-bit dynamically linked shared library x86_64
project-123df13e-01929daa.eln: Mach-O 64-bit dynamically linked shared library x86_64
projectile-5f2672ce-c7829720.eln: Mach-O 64-bit dynamically linked shared library x86_64
s-8e2025a1-61be6d57.eln: Mach-O 64-bit dynamically linked shared library x86_64
shrink-path-0eeb151b-7aa5218d.eln: Mach-O 64-bit dynamically linked shared library x86_64
smartparens-8d4e76ff-c75f7f23.eln: Mach-O 64-bit dynamically linked shared library x86_64
smartparens-config-087fcbf2-ee658dc3.eln: Mach-O 64-bit dynamically linked shared library x86_64
smartparens-html-60d88526-30aaa416.eln: Mach-O 64-bit dynamically linked shared library x86_64
smartparens-text-b232a5f9-309c6a03.eln: Mach-O 64-bit dynamically linked shared library x86_64
solaire-mode-76885170-83bf4e31.eln: Mach-O 64-bit dynamically linked shared library x86_64
subr--trampoline-64656c6574652d66696c65_delete_file_0.eln: Mach-O 64-bit dynamically linked shared library x86_64
subr--trampoline-7365742d6275666665722d6d756c746962797465_set_buffer_multibyte_0.eln: Mach-O 64-bit dynamically linked shared library x86_64
subr--trampoline-646566696e652d6b6579_define_key_0.eln: Mach-O 64-bit dynamically linked shared library x86_64
transient-acbbb9d3-7d7f170b.eln: Mach-O 64-bit dynamically linked shared library x86_64
undo-fu-adf5a5dc-5e9aa576.eln: Mach-O 64-bit dynamically linked shared library x86_64
undo-fu-session-f5ac9d8e-fa9858bd.eln: Mach-O 64-bit dynamically linked shared library x86_64
use-package-core-38e7bdb9-a8e5f021.eln: Mach-O 64-bit dynamically linked shared library arm64
vertico-3b597395-59a88888.eln: Mach-O 64-bit dynamically linked shared library x86_64
vertico-directory-a794f120-578b8bf5.eln: Mach-O 64-bit dynamically linked shared library x86_64
vertico-repeat-a8ec7323-a64d2810.eln: Mach-O 64-bit dynamically linked shared library x86_64
wakatime-mode-5022c804-183fb426.eln: Mach-O 64-bit dynamically linked shared library x86_64
web-mode-7f360c42-2d655d7e.eln: Mach-O 64-bit dynamically linked shared library x86_64
which-key-eba58ea2-1f3e0aa8.eln: Mach-O 64-bit dynamically linked shared library x86_64
ws-butler-ef2120db-9b48b92a.eln: Mach-O 64-bit dynamically linked shared library x86_64
yasnippet-c25cab0a-8df7bbfb.eln: Mach-O 64-bit dynamically linked shared library x86_64
```

It maybe not a problem in doomemacs, but I haven't found the root cause, please give some direction if you find any details, thanks❤️ 

### Steps to reproduce

To reproduce the problem, you need to:
1. Use Apple M1 (M2 is not tested).
2. Use an x86 Emacs.
3. Use Nix.

Steps:
1. Install doomemacs.
2. Clone [this repository](https://github.com/goofansu/emacs-nativecomp-error-in-direnv) and run `direnv allow`.
3. Open Emacs.app and `find-file` to `emacs-nativecomp-error-in-direnv`.
4. Native compilation generates some eln files in arm64 architecture.

### System Information

https://gist.github.com/goofansu/bc7ce1a5963daed46e0fe6cc752c37e3

**Comment by BuddhiLW** [comment]
```quote
### What Did We Just Learn?[^1]

1. This environment can be complicated and hard to grok.
2. There are tools which can help, some of which are MacOS specific, so you may not even know they exist if you’re used to a Linux environment.
3. There are some horrible hacks (such as re-execing a shell) which can help too.
4. You will need to modify your environment (and in particular the directories searched in your PATH) to make things work well.
5. You can get this all to work.
6. It has got much easier now that brew has AArch64 MacOS support.
```

[^1]: [https://cpufun.substack.com/p/setting-up-the-apple-m1-for-native](https://cpufun.substack.com/p/setting-up-the-apple-m1-for-native)

**Comment by BuddhiLW** [comment]
No problem; happy to help :D

**Comment by goofansu** [comment]
Skipped the problem by installing [railwaycat/emacsmacport](https://github.com/railwaycat/emacsmacport).

```shell
brew tap railwaycat/emacsmacport
brew install --cask emacs-mac
```

Notice the cask installed version doesn't enable the native compilation.

First I installed it from source by running `brew install emacs-mac --with-native-compilation`, but it requires to create a soft link to the `/Applications` folder. It has some cons, for example, [Alfred](https://alfred.app/) cannot search the results of soft links, so finally I installed the cask version.
